### ISSUE TITLE
LoadBalancer type service load balance traffic on pods in different namespace 

### ISSUE THREAD

**Issue by rvrignaud** [kubernetes/kubernetes#22456]
Hello,

I use a GKE cluster (master and nodes in 1.1.8 version). I'm using the iptables mode for kube-proxy.
In this cluster I have multiple namespaces.
In one of the namespace I have a service defined as:
```
{
  "kind": "Service",
  "apiVersion": "v1",
  "metadata": {
    "name": "frontend",
    "namespace": "int",
    "labels": {
       "name": "frontend"
    }
  },
  "spec": {
    "type": "LoadBalancer",
    "loadBalancerIP": "$PUBLIC_IP",
    "ports": [
      {
        "protocol": "TCP",
        "port": 80,
        "targetPort": 80,
        "name": "http"
      },
      {
        "protocol": "TCP",
        "port": 443,
        "targetPort": 443,
        "name": "https"
      }
    ],
    "selector": {
      "name": "frontend"
    }
  }
}
```

When I use the public IP of this service, requests are loadbalanced between pods in 2 different namespaces (int and int2) that matches the selector "frontend".
In the "int" namespace I have the following error in events:
```
FIRSTSEEN   LASTSEEN   COUNT     NAME       KIND      SUBOBJECT   REASON                       SOURCE                  MESSAGE
1d          5s         3905      frontend   Service               CreatingLoadBalancer         {service-controller }   Creating load balancer
9h          5s         3898      frontend   Service               CreatingLoadBalancerFailed   {service-controller }   Error creating load balancer (will retry): Failed to create load balancer for service int/frontend: failed to create forwarding rule a15edb77cc28911e591d142010af000b: googleapi: Error 400: Invalid value for field 'resource': ''.  Specified IP address is in-use and would result in a conflict., invalid
```

When I use the CLUSTER_IP address directly, all the requests are going to the pod in the same namespace so the problem is limited to the external loadbalancer.

The problem seemed to start when I played in changing selector name and applying configuration with:
```
kubectl --namespace=int apply -f file.json
```

Here are some details about google cloud forward-rules:
```
gcloud compute forwarding-rules list | grep $PUBLIC_IP
a387705ded6ee11e5819142010af000b europe-west1 $PUBLIC_IP   TCP         europe-west1/targetPools/a387705ded6ee11e5819142010af000b

gcloud compute target-pools describe a387705ded6ee11e5819142010af000b --region europe-west1
creationTimestamp: '2016-03-03T02:35:00.413-08:00'
description: ''
id: '4525248592704023995'
instances:
- https://www.googleapis.com/compute/v1/projects/clustree-cloud/zones/europe-west1-c/instances/gke-stg-b9e96ddd-node-ojvs
- https://www.googleapis.com/compute/v1/projects/clustree-cloud/zones/europe-west1-c/instances/gke-stg-b9e96ddd-node-rkkp
- https://www.googleapis.com/compute/v1/projects/clustree-cloud/zones/europe-west1-c/instances/gke-stg-b9e96ddd-node-j5fm
- https://www.googleapis.com/compute/v1/projects/clustree-cloud/zones/europe-west1-c/instances/gke-stg-b9e96ddd-node-uqsv
- https://www.googleapis.com/compute/v1/projects/clustree-cloud/zones/europe-west1-c/instances/gke-stg-b9e96ddd-node-l1lg
- https://www.googleapis.com/compute/v1/projects/clustree-cloud/zones/europe-west1-c/instances/gke-stg-b9e96ddd-node-0a99
- https://www.googleapis.com/compute/v1/projects/clustree-cloud/zones/europe-west1-c/instances/gke-stg-b9e96ddd-node-9wfj
- https://www.googleapis.com/compute/v1/projects/clustree-cloud/zones/europe-west1-c/instances/gke-stg-b9e96ddd-node-al0f
- https://www.googleapis.com/compute/v1/projects/clustree-cloud/zones/europe-west1-c/instances/gke-stg-b9e96ddd-node-e1f7
- https://www.googleapis.com/compute/v1/projects/clustree-cloud/zones/europe-west1-c/instances/gke-stg-b9e96ddd-node-060z
- https://www.googleapis.com/compute/v1/projects/clustree-cloud/zones/europe-west1-c/instances/gke-stg-b9e96ddd-node-8vs9
- https://www.googleapis.com/compute/v1/projects/clustree-cloud/zones/europe-west1-c/instances/gke-stg-b9e96ddd-node-dxgy
- https://www.googleapis.com/compute/v1/projects/clustree-cloud/zones/europe-west1-c/instances/gke-stg-b9e96ddd-node-nf8r
- https://www.googleapis.com/compute/v1/projects/clustree-cloud/zones/europe-west1-c/instances/gke-stg-b9e96ddd-node-lrb5
- https://www.googleapis.com/compute/v1/projects/clustree-cloud/zones/europe-west1-c/instances/gke-stg-b9e96ddd-node-wunh
kind: compute#targetPool
name: a387705ded6ee11e5819142010af000b
region: https://www.googleapis.com/compute/v1/projects/clustree-cloud/regions/europe-west1
selfLink: https://www.googleapis.com/compute/v1/projects/clustree-cloud/regions/europe-west1/targetPools/a387705ded6ee11e5819142010af000b
```

This problem seems limited to this only service.
I'm not sure how to debug more this issue and what information could be added to help diagnostic the problem.

Thanks for any help.

**Comment by thockin** [comment]
> When I use the public IP of this service, requests are loadbalanced between pods in 2 different namespaces (int and int2) that matches the selector 
> "frontend".

That should not be.  Services do not cross namespaces.  I can not reproduce this behavior at all - are you sure this is what you're seeing?

Can you send me `iptables-save` from a node?

**Comment by rvrignaud** [comment]
> That should not be. Services do not cross namespaces. I can not reproduce this behavior at all - are you sure this is what you're seeing?

I'm sure. This have been working well for me for months.
I'm sending you the iptables-save privately.

**Comment by thockin** [comment]
The issue is that the same LB IP was used for two services.  That hole is already closed in 1.2, and not worth a 1.1.x patch on its own.  This also exposes an interesting side-effect of the impl: since we iterate a map to populate iptables, and we know Go maps are not order-stable, the iptables on different nodes comes out in different orders.  I will contemplate whether that is a "bug" or just "interesting" (10 bucks says it blows up for someone, now that we know about it).

Load-balancing across namespaces is not an intentional thing, and is not a feature of the system :)
